src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +249,522 @@ otError DuaManager::Store(void)
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+
+    VerifyOrExit(mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1(), OT_NOOP);
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+
+    // Prepare DUA.req
+    VerifyOrExit((message = Get<Coap::Coap>().NewPriorityMessage()) != nullptr, error = OT_ERROR_NO_BUFS);
+
+    SuccessOrExit(error =
+                      message->Init(OT_COAP_TYPE_CONFIRMABLE, OT_COAP_CODE_POST, OT_URI_PATH_DUA_REGISTRATION_REQUEST));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (mDuaState == kToRegister && mDelay.mFields.mRegistrationDelay == 0)
+    {
+        dua = GetDomainUnicastAddress();
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kTarget, &dua, sizeof(dua)));
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kMeshLocalEid, &mle.GetMeshLocal64().GetIid(),
+                                             sizeof(Ip6::InterfaceIdentifier)));
+        mDuaState             = kRegistering;
+        mLastRegistrationTime = TimerMilli::GetNow();
+    }
+    else
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+    {
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        if (!mRegisterCurrentChildIndex)
+        {
+            for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValid))
+            {
+                uint16_t childIndex = Get<ChildTable>().GetChildIndex(child);
+
+                if (mChildDuaMask.Get(childIndex) && !mChildDuaRegisterMask.Get(childIndex))
+                {
+                    mChildIndexDuaRegistering = childIndex;
+                }
+            }
+        }
+        {
+            uint32_t lastTransactionTime;
+            Child *  child = Get<ChildTable>().GetChildAtIndex(mChildIndexDuaRegistering);
+
+            OT_ASSERT(child->GetDomainUnicastAddress(dua) == OT_ERROR_NONE);
+
+            lastTransactionTime = Time::MsecToSec(TimerMilli::GetNow() - child->GetLastHeard());
+
+            SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kTarget, &dua, sizeof(dua)));
+            SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kMeshLocalEid, &child->GetMeshLocalIid(),
+                                                 sizeof(Ip6::InterfaceIdentifier)));
+            SuccessOrExit(error = Tlv::AppendUint32Tlv(*message, ThreadTlv::kLastTransactionTime, lastTransactionTime));
+        }
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    }
+
+    if (!mle.IsFullThreadDevice() && mle.GetParent().IsThreadVersion1p1())
+    {
+        uint8_t pbbrServiceId;
+
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetServiceId(pbbrServiceId));
+        SuccessOrExit(error = mle.GetServiceAloc(pbbrServiceId, messageInfo.GetPeerAddr()));
+    }
+    else
+    {
+        messageInfo.GetPeerAddr().SetToRoutingLocator(mle.GetMeshLocalPrefix(),
+                                                      Get<BackboneRouter::Leader>().GetServer16());
+    }
```

Comment:
This logic is used by `SendMulticastListenerRegistration` too. Should we use a common method to encapsulate this logic?

Reply:
Fine to me.